### DISCUSSION TITLE
Array.prototype and Object.prototype for in and for each loop issue

### DISCUSSION THREAD

**Discussion by MalacTheLittle** [airsdk/Adobe-Runtime-Support#3272] [Q&A]
If I extend `Array` or `Object` with `Array.prototype` custom methods, `for each` and `for in` loop are reaching prototype functions, e.g.:
```
Array.prototype._last = function():*{
   return this[this.length-1];
}

var bar:Array = ["a","b","c"];

for(var foo in bar){
   trace(foo);
}
```
Trace result:
```
a
b
c
_last
```

Is there a way to extend Array/Object to avoid such for loop behaviour?

**Comment by ylazy** [comment]
Using prototype to extend a type is not recommended. I think there's no way to exclude prototype methods/properties from for loop because Array/Object is dynamic types, so anything you define through prototype is enumerable. That's why we often use a static class (utility class) to do this, or, extend them with a sub-class.

**Reply by ajwfrost** [reply]
I find this kind of thing quite fascinating :-)

A quick look at the generated ABC and it's basically calling "hasnext2" and finding that there's another property, and "nextname" to find its value.

The documentation for "hasnext2":
> Get the index of the next property after the property located at index _cur_index_ on object _obj_. If there are no more properties on _obj_, then _obj_ is set to the next object on the prototype chain of _obj_ ....

So it's basically doing what it's meant to.

I would agree perhaps extending Array is a way to provide this function? but you would have more overhead for that, as 'array' has built-in ABC instructions whereas your subclass would not. So a utility class may be more efficient overall.

Interesting how Array is defined: https://github.com/adobe/avmplus/blob/master/core/Array.as
I'd wondered whether you might be able to work around it using some namespaces.. but ultimately, whatever is in these source files gets preprocessed and turned into the traits of the built-in type, and adding things dynamically at runtime is going to be treated differently..

thanks

**Reply by MalacTheLittle** [reply]
There are some AS3 "syntax lacks" that would be realy handy from readability point of view, like Array.last(), Array.last(-1), Array.first(), Array.has(), Object.removeChild(), Date.getWeek()...
In order to convey it more closely, reaching last record in long named variables:
`iNeedToHaveLongNamedArrayVariableButItLooksLikeSausage[iNeedToHaveLongNamedArrayVariableButItLooksLikeSausage.length-1]` 
I don't need to mention making Math operations with it...

So, all of this could be easily solved if prototyping of top level classes wouldn't affect it's basic operability.

I know the best solution is to ask Harman to update such "lacks", but for most non-urgent updates it takes a long time (usually years) to be implemented (which is also understandable since they deal with the problems of changing operating systems on a daily basis), so having a way to make syntax personalization would be more than welcomed. 

> I would agree perhaps extending Array is a way to provide this function? but you would have more overhead for that, as 'array' has built-in ABC instructions whereas your subclass would not. So a utility class may be more efficient overall.

I was thinking of extending Array, but then I need to declare it with `new`, can't use `var a = []`.

**Reply by ajwfrost** [reply]
One of the key reasons for limiting the AS3 API updates is to try to keep all the numbers in sync i.e. when the ActionScript definitions change, there needs to be a new SWF version number and a new namespace number.. when we started out, we had been changing the ActionScript APIs without updating the SWF version which means it's very possible to create a SWF with one version of AIR SDK 33.1 that wouldn't run on the AIR 33.1 runtime from an earlier version.. which is bad, generally we would want these incompatibilities to be flagged up at compile-time. So, you build against the "airglobal.swc" definition library, and there shouldn't ever be an issue...

But anyway - yes I take your point that it's quicker to then create your own variants and extensions of the classes as you need.

For the modification of prototypes though, this is how it's meant to work... there's a similar discussion on it here:
https://stackoverflow.com/questions/11337849/ways-to-extend-array-object-in-javascript

Someone makes the point that the "foreach()" method will break if you extend the prototype, and all of the suggestions that they're making in the comments do result in that new method appearing if you do "for x in y" ... so you'd likely need to define a new method separately for the iteration and just use that instead. But yes, either way, you'll end up not using some of the built-in language accelerators for array (e.g. in the compiler such as `var a = [ ]` if you subclass vs `for (x in y)` if you extend the prototype; and in the ABC format itself with instructions such as `newarray` and `nextname` etc)

Not sure there's an easy solution for this. Probably the most efficient way to do it would be to contain all the extra functionality into a new utility class, so e.g. you can call `Utilities.ArrayLast(array)` to get the last value, and `Utilities.DateGetWeek(date)` etc for some of your other ideas. This at least keeps the code separate and means the compiler/assembled ActionScript is as efficient as possible, and the functions perhaps be more easily swapped out via Find/Replace as we get round to the AS3 API updates..?

thanks

**Reply by MalacTheLittle** [reply]
Right now I'm actually using a separate class for all my shorthands, but it would be more convenient and efficient as `Array.last()`. 
Actually, I discovered this issue when I prototyped Array and it "broke" my shorthands class.

Anyway, keep up the good work and don't forget all those good ideas in section #2152 :)
